pkg/extensions/sync/sync.go

```diff
@@ -0,0 +1,436 @@
+
+	for _, repo := range repos {
+		for _, c := range content {
+			if strings.HasPrefix(repo, c.Prefix) {
+				filtered = append(filtered, repo)
+			}
+		}
+	}
+
+	return filtered
+}
+
+// parseRepositoryReference parses input into a reference.Named, and verifies that it names a repository, not an image.
+func parseRepositoryReference(input string) (reference.Named, error) {
+	ref, err := reference.ParseNormalizedNamed(input)
+	if err != nil {
+		return nil, err
+	}
+
+	if !reference.IsNameOnly(ref) {
+		return nil, errors.ErrInvalidRepositoryName
+	}
+
+	return ref, nil
+}
+
+// getImageTags lists all tags in a repository.
+// It returns a string slice of tags and any error encountered.
+func getImageTags(ctx context.Context, sysCtx *types.SystemContext, repoRef reference.Named) ([]string, error) {
+	dockerRef, err := docker.NewReference(reference.TagNameOnly(repoRef))
+	if err != nil {
+		return nil, err // Should never happen for a reference with tag and no digest
+	}
+
+	tags, err := docker.GetRepositoryTags(ctx, sysCtx, dockerRef)
+	if err != nil {
+		return nil, err
+	}
+
+	return tags, nil
+}
+
+// getTagFromRef returns a tagged reference from an image reference.
+func getTagFromRef(ref types.ImageReference, log log.Logger) reference.Tagged {
+	tagged, isTagged := ref.DockerReference().(reference.Tagged)
+	if !isTagged {
+		log.Warn().Msgf("Internal error, reference %s does not have a tag, skipping", ref.DockerReference())
+		return nil
+	}
+
+	return tagged
+}
+
+// filterImagesByTagRegex filters images by tag regex give in the config.
+func filterImagesByTagRegex(sourceReferences *[]types.ImageReference, content []Content, log log.Logger) error {
+	log.Info().Msgf("Start filtering using the regular expression: %#v", content)
+
+	refs := *sourceReferences
+
+	for _, c := range content {
+		if c.Tags == nil {
+			continue
+		}
+
+		if c.Tags.Regex != nil {
+			tagReg, err := regexp.Compile(*c.Tags.Regex)
+			if err != nil {
+				return err
+			}
+
+			for i := 0; i < len(refs); i++ {
+				tagged := getTagFromRef(refs[i], log)
+				if tagged == nil {
+					continue
+				}
+
+				if !tagReg.MatchString(tagged.Tag()) {
+					refs = append(refs[:i], refs[i+1:]...)
+					i--
+				}
+			}
+		}
+	}
+
+	*sourceReferences = refs
+
+	return nil
+}
+
+// filterImagesBySemver filters images by checking if their tags are semver compliant.
+func filterImagesBySemver(sourceReferences *[]types.ImageReference, content []Content, log log.Logger) {
+	log.Info().Msg("Start filtering using semver compliant rule")
+
+	refs := *sourceReferences
+
+	for _, c := range content {
+		if c.Tags == nil {
+			continue
+		}
+
+		if c.Tags.Semver != nil {
+			for i := 0; i < len(refs); i++ {
+				tagged := getTagFromRef(refs[i], log)
+				if tagged == nil {
+					continue
+				}
+
+				_, ok := semver.NewVersion(tagged.Tag())
+				if ok != nil {
+					log.Info().Msgf("tag %s is not semver, skipping", tagged.Tag())
+
+					refs = append(refs[:i], refs[i+1:]...)
```

Comment:
I'm deleting refs (inplace) which are not semver compliant.
Because my for loops until "i"<len(refs) I need to decrement "i" when I'm deleting a ref. Otherwise  this line:
tagged := getTagFromRef(refs[i], log)
will get index error.

Reply:
and refs[i+1] will never give out of range ?

Reply:
https://play.golang.org/p/WUfd-ENZfFT

I played with the code, I guess I'm not missing anything.

Reply:
Note in your example you always skip the first element.

An other interesting perspective here: https://zetcode.com/golang/filter-slice/